nftables.go

```diff
@@ -298,71 +295,169 @@ func (n *nft) Init() error {
-			return nil
-		}
-		if err := n.conn6.SetDeleteElements(n.set6, []nftables.SetElement{{Key: []byte(net.ParseIP(*decision.Value).To16())}}); err != nil {
-			return err
-		}
-		if err := n.conn6.Flush(); err != nil {
-			return err
+func (n *nft) commitAddedDecisions() error {
+	n.decisionsToAdd = normalizedDecisions(n.decisionsToAdd)
+	addedIPV6, addedIPV4 := false, false
+	currentState := make(map[string]struct{})
+	var err error
+	for i := 0; i < len(n.decisionsToAdd); {
+		if i == 0 || addedIPV4 || addedIPV6 {
+			currentState, err = n.getCurrentState()
+			if err != nil {
+				return err
+			}
+			addedIPV4, addedIPV6 = false, false
 		}
-	} else { // ipv4
-		if n.conn != nil {
-			var ipAddr string
-			if strings.Contains(*decision.Value, "/") {
-				ipAddr = strings.Split(*decision.Value, "/")[0]
+		for canAdd := 200; canAdd > 0 && i < len(n.decisionsToAdd); {
+			decision := n.decisionsToAdd[i]
+			i++
+			decisionIP := net.ParseIP(*decision.Value)
+			if _, ok := currentState[decisionIP.String()]; ok {
+				log.Debugf("skipping %s since it's already in set", decisionIP)
+
 			} else {
-				ipAddr = *decision.Value
+				if strings.Contains(decisionIP.String(), ":") && n.conn6 != nil {
+					if err := n.conn6.SetAddElements(n.set6, []nftables.SetElement{{Key: decisionIP.To16()}}); err != nil {
+						return err
+					}
+					addedIPV6 = true
+				} else {
+					if err := n.conn.SetAddElements(n.set, []nftables.SetElement{{Key: decisionIP.To4()}}); err != nil {
+						return err
+					}
+					addedIPV4 = true
+				}
+				canAdd--
+				log.Debugf("adding %s to buffer ", decisionIP)
 			}
-			if err := n.conn.SetDeleteElements(n.set, []nftables.SetElement{{Key: net.ParseIP(ipAddr).To4()}}); err != nil {
+		}
+		if addedIPV4 {
+			if err := n.conn.Flush(); err != nil {
 				return err
 			}
-			if err := n.conn.Flush(); err != nil {
+		}
+		if addedIPV6 {
+			if err := n.conn6.Flush(); err != nil {
 				return err
 			}
 		}
 	}
+	return nil
+}
 
+func (n *nft) Commit() error {
+	defer n.reset()
+	if err := n.commitDeletedDecisions(); err != nil {
+		return err
+	}
+	if err := n.commitAddedDecisions(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func elementSliceToIPSet(elements []nftables.SetElement) map[string]struct{} {
+	ipSet := make(map[string]struct{})
+	for _, element := range elements {
+		ipSet[net.IP(element.Key).String()] = struct{}{}
+	}
+	return ipSet
+}
+
+// remove duplicates, normalize decision timeouts
+func normalizedDecisions(decisions []*models.Decision) []*models.Decision {
+	vals := make(map[string]struct{})
+	finalDecisions := make([]*models.Decision, 0)
+	for _, d := range decisions {
+		if _, ok := vals[*d.Value]; ok {
+			continue
+		}
+		vals[*d.Value] = struct{}{}
+		if _, err := time.ParseDuration(*d.Duration); err != nil {
+			d.Duration = &defaultTimeout
+		}
+		*d.Value = strings.Split(*d.Value, "/")[0]
+		finalDecisions = append(finalDecisions, d)
+	}
```

Comment:
Hi @sbs2001, now the logic is that the first decision will be saved, should it actually be the latest decision that gets a place in the normalized array?